proto/runner.proto

```diff
@@ -17,10 +17,9 @@ message RunRequest {
   // A minimum of commit_sha or branch must be specified.
   git.RepoState repo_state = 3;
 
-  // The bazel command to run. Required.
-  // Example: "build :server"
-  // Example: "test //..."
-  string bazel_command = 4;
+  // The commands to run, in order. Required.
+  // Example: ["bazel build :server", "echo \"Hello World\""]
+  repeated string steps = 16;
```

Comment:
Having a repeated list of strings feels pretty similar to having a single string, since each script can have multiple commands (joined by `;` / `&&` etc.)

To make this a little easier in the common case, we could make it a single string called `run` (i.e. one step, and called "run" to match the workflow config YAML)

Alternatively, if we want to go for maximum extensibility, and allow adding more features like `name`, `if`, `env`, `uses`, etc., we could make it a repeated struct like `"steps": [ { "name": "...", "run": "..."} ]` (i.e. mirror the workflow config YAML)

wdyt?

Reply:
I'm torn but I'm leaning towards the more extensible approach. The biggest reason is that I think we would want this as our CI product matures down the line, and we wouldn't have to require a migration

It also gives us more flexibility in making a better product experience. One feature I expect would be popular is having a `timeout` on certain steps, for example

That being said, we're trying not to prioritize a comprehensive CI suite right now, and that decision would open us to more customers asking us to add special features as opposed to if we were just like "do it all in bash". Plus each command is run in a separate bash process, which might lead to surprising behavior vs if we just ran it all in one block

CC @siggisim for additional thoughts

Reply:
Siggi and I talked offline. We agreed that the struct approach is okay because while it's more complex, this is intended as an internal API. We can make an external API with a simpler interface, and extend the remote bazel cli so that fewer customers need to use the internal API

An idea I had is that we can also make a cli command that outputs a curl command with a sample Remote Run API call. So if you need to tweak one thing, you don't have to manually create a complicated request